pilot/pkg/proxy/envoy/v2/xds.go

```diff
@@ -24,75 +27,135 @@ import (
 	}
-	out.snapshots = cache.NewSnapshotCache(false, out, nil)
+	out.snapshots = cache.NewSnapshotCache(false /* non-ADS mode */, out, nil /* add spam logger here */)
 	out.server = server.NewServer(out.snapshots, out)
 	return out
 }
 
-// Register with gRPC server
-func (cache *ConfigCache) Register(grpcServer *grpc.Server) {
-	v2.RegisterRouteDiscoveryServiceServer(grpcServer, cache.server)
-	// v2.RegisterClusterDiscoveryServiceServer(grpcServer, cache.server)
-	// v2.RegisterListenerDiscoveryServiceServer(grpcServer, cache.server)
+// RegisterOutput registers with a gRPC server
+// TODO: make sure the inputs are initialized before serving
+func (xds *ConfigCache) RegisterOutput(grpcServer *grpc.Server) {
+	v2.RegisterRouteDiscoveryServiceServer(grpcServer, xds.server)
 }
 
 // RegisterInput connects with the controllers
-func (cache *ConfigCache) RegisterInput(services model.Controller, configs model.ConfigStoreCache) {
-	if err := services.AppendServiceHandler(cache.OnServiceEvent); err != nil {
+// The route generation code needs to subscribe to:
+// - collection of services
+// - collection of virtual services
+func (xds *ConfigCache) RegisterInput(services model.Controller, configs model.ConfigStoreCache) {
+	if err := services.AppendServiceHandler(xds.OnServiceEvent); err != nil {
 		log.Fatal(err)
 	}
-	configs.RegisterEventHandler(model.VirtualService.Type, cache.OnConfigEvent)
-	configs.RegisterEventHandler(model.DestinationRule.Type, cache.OnConfigEvent)
+	configs.RegisterEventHandler(model.VirtualService.Type, xds.OnConfigEvent)
 }
 
-// ID ...
-func (cache *ConfigCache) ID(node *core.Node) string {
+// ID uses string-encoded node ID
+func (xds *ConfigCache) ID(node *core.Node) string {
 	return node.GetId()
 }
 
 // OnServiceEvent ...
-func (cache *ConfigCache) OnServiceEvent(svc *model.Service, event model.Event) {
-	// TODO
+func (xds *ConfigCache) OnServiceEvent(svc *model.Service, event model.Event) {
+	xds.Update()
 }
 
 // OnConfigEvent ...
-func (cache *ConfigCache) OnConfigEvent(svc model.Config, event model.Event) {
-	// TODO
+func (xds *ConfigCache) OnConfigEvent(svc model.Config, event model.Event) {
+	xds.Update()
 }
 
-// OnStreamOpen ...
-func (cache *ConfigCache) OnStreamOpen(int64, string) {}
+// Update performs the global re-compute
+// TODO make this more incremental
+func (xds *ConfigCache) Update() {
+	xds.mu.Lock()
+	defer xds.mu.Unlock()
+
+	xds.version++
+	services, _ := xds.services.Services()
+	virtualServices, _ := xds.configs.List(model.VirtualService.Type, model.NamespaceAll)
+	xds.hosts = v1alpha3.TranslateVirtualHosts(virtualServices, v1alpha3.IndexServices(services), xds.serviceSuffix)
+	for id := range xds.nodes {
+		xds.updateID(id)
+	}
+}
 
-// OnStreamClosed ...
-func (cache *ConfigCache) OnStreamClosed(int64) {}
+func (xds *ConfigCache) updateID(id string) {
+	proxy, _ := model.ParseServiceNode(id)
+	switch proxy.Type {
+	case model.Ingress: // TODO ingress nodes are deprecated?
```

Comment:
No, they are not.

Reply:
actually, model.Ingress is, but not model.Router. Its a small change in the YAML template to update the ingress proxy's node ID.